File: clients/libs/rust/src/lightning_latch.rs

use crate::{client_config::ClientConfig, sqlite_manager::get_wallet};
use anyhow::{anyhow, Result};
use mercurylib::transfer::sender::{PaymentHashRequestPayload, PaymentHashResponsePayload};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct CreatePreImageResponse {
    pub pre_image: String,
    pub batch_id: String,
}

pub async fn create_pre_image(
    client_config: &ClientConfig, 
    wallet_name: &str, 
    statechain_id: &str) -> Result<CreatePreImageResponse> 
{
    let batch_id = Some(uuid::Uuid::new_v4().to_string()).unwrap();

    let mut wallet: mercurylib::wallet::Wallet = get_wallet(&client_config.pool, &wallet_name).await?;

    let coin = wallet.coins
        .iter_mut()
        .filter(|tx| tx.statechain_id == Some(statechain_id.to_string())) // Filter coins with the specified statechain_id
        .min_by_key(|tx| tx.locktime.unwrap_or(u32::MAX)); // Find the one with the lowest locktime

    if coin.is_none() {
        return Err(anyhow!("No coins associated with this statechain ID were found"));
    }

    let coin = coin.unwrap();

    let signed_statechain_id = coin.signed_statechain_id.as_ref().unwrap();

    let payment_hash_payload = PaymentHashRequestPayload {
        statechain_id: statechain_id.to_string(),
        auth_sig: signed_statechain_id.to_string(),
        batch_id: batch_id.clone(),
    };

    let endpoint = client_config.statechain_entity.clone();
    let path = "transfer/paymenthash";

    let client = client_config.get_reqwest_client()?;
    let request = client.post(&format!("{}/{}", endpoint, path));

    let response = request.json(&payment_hash_payload).send().await?;

    if response.status() != 200 {
        let response_body = response.text().await?;
        return Err(anyhow!(response_body));
    }

    let value = response.text().await?;

    let payment_hash_response_payload: PaymentHashResponsePayload = serde_json::from_str(value.as_str())?;

    Ok(CreatePreImageResponse {
        pre_image: payment_hash_response_payload.hash,
        batch_id,
    })
}

pub async fn confirm_pending_invoice(client_config: &ClientConfig, wallet_name: &str, statechain_id: &str) -> Result<()> {

    let mut wallet: mercurylib::wallet::Wallet = get_wallet(&client_config.pool, &wallet_name).await?;

    let coin = wallet.coins
        .iter_mut()
        .filter(|tx| tx.statechain_id == Some(statechain_id.to_string())) // Filter coins with the specified statechain_id
        .min_by_key(|tx| tx.locktime.unwrap_or(u32::MAX)); // Find the one with the lowest locktime

    if coin.is_none() {
        return Err(anyhow!("No coins associated with this statechain ID were found"));
    }

    let coin = coin.unwrap();

    let signed_statechain_id = coin.signed_statechain_id.as_ref().unwrap();

    let path = "transfer/unlock";

    let client = client_config.get_reqwest_client()?;
    let request = client.post(&format!("{}/{}", client_config.statechain_entity, path));

    let transfer_unlock_request_payload = mercurylib::transfer::receiver::TransferUnlockRequestPayload {
        statechain_id: statechain_id.to_string(),
        auth_sig: signed_statechain_id.to_string(),
        auth_pub_key: None,
    };

    let status = request.json(&transfer_unlock_request_payload).send().await?.status();

    if !status.is_success() {
        return Err(anyhow::anyhow!("Failed to update transfer message".to_string()));
    }

    Ok(())
}

File: clients/libs/nodejs/lightning-latch.js
const sqlite_manager = require('./sqlite_manager');
const { v4: uuidv4 } = require('uuid');
const axios = require('axios').default;
const { SocksProxyAgent } = require('socks-proxy-agent');
const { CoinStatus } = require('./coin_enum');

const createPreImage  = async (clientConfig, db, walletName, statechainId) => {

    const batchId = uuidv4();

    let wallet = await sqlite_manager.getWallet(db, walletName);

    let coinsWithStatechainId = wallet.coins.filter(c => {
        return c.statechain_id === statechainId
    });

    if (!coinsWithStatechainId || coinsWithStatechainId.length === 0) {
        throw new Error(`There is no coin for the statechain id ${statechainId}`);
    }

    // If the user sends to himself, he will have two coins with same statechain_id
    // In this case, we need to find the one with the lowest locktime
    // Sort the coins by locktime in ascending order and pick the first one
    let coin = coinsWithStatechainId.sort((a, b) => a.locktime - b.locktime)[0];

    if (coin.status != CoinStatus.CONFIRMED && coin.status != CoinStatus.IN_TRANSFER) {
        throw new Error(`Coin status must be CONFIRMED or IN_TRANSFER to transfer it. The current status is ${coin.status}`);
    }

    if (coin.locktime == null) {
        throw new Error("Coin.locktime is null");
    }

    let paymentHashPayload  = {
        statechain_id: statechainId,
        auth_sig: coin.signed_statechain_id,
        batch_id: batchId
    };

    let paymentHash = await sendPaymentHash(clientConfig, paymentHashPayload);

    return {
        hash: paymentHash,
        batchId: batchId
    };
}

const sendPaymentHash = async (clientConfig, paymentHashPayload) => {

    const url = `${clientConfig.statechainEntity}/transfer/paymenthash`;
    const torProxy = clientConfig.torProxy;

    let socksAgent = undefined;

    if (torProxy) {
        socksAgent = { httpAgent: new SocksProxyAgent(torProxy) };
    }
    
    let response = await axios.post(url, paymentHashPayload, socksAgent);

    return response?.data?.hash;
}

const confirmPendingInvoice = async (clientConfig, db, walletName, statechainId) => {

    let wallet = await sqlite_manager.getWallet(db, walletName);

    let coinsWithStatechainId = wallet.coins.filter(c => {
        return c.statechain_id === statechainId
    });

    if (!coinsWithStatechainId || coinsWithStatechainId.length === 0) {
        throw new Error(`There is no coin for the statechain id ${statechainId}`);
    }

    // If the user sends to himself, he will have two coins with same statechain_id
    // In this case, we need to find the one with the lowest locktime
    // Sort the coins by locktime in ascending order and pick the first one
    let coin = coinsWithStatechainId.sort((a, b) => a.locktime - b.locktime)[0];

    const transferUnlockRequestPayload = {
        statechain_id: statechainId,
        auth_sig: coin.signed_statechain_id,
        auth_pub_key: null
    };

    const url = `${clientConfig.statechainEntity}/transfer/unlock`;
    const torProxy = clientConfig.torProxy;

    let socksAgent = undefined;

    if (torProxy) {
        socksAgent = { httpAgent: new SocksProxyAgent(torProxy) };
    }
    
    // If there is an http error an exception will be thrown
    await axios.post(url, transferUnlockRequestPayload, socksAgent);
}

const retrievePreImage = async (clientConfig, db, walletName, statechainId, batchId) => {

    let wallet = await sqlite_manager.getWallet(db, walletName);

    let coinsWithStatechainId = wallet.coins.filter(c => {
        return c.statechain_id === statechainId
    });

    if (!coinsWithStatechainId || coinsWithStatechainId.length === 0) {
        throw new Error(`There is no coin for the statechain id ${statechainId}`);
    }

    // If the user sends to himself, he will have two coins with same statechain_id
    // In this case, we need to find the one with the lowest locktime
    // Sort the coins by locktime in ascending order and pick the first one
    let coin = coinsWithStatechainId.sort((a, b) => a.locktime - b.locktime)[0];

    const transferPreimageRequestPayload = {
        statechain_id: statechainId,
        auth_sig: coin.signed_statechain_id,
        previous_user_auth_key: coin.auth_pubkey,
        batch_id: batchId,
    };

    const url = `${clientConfig.statechainEntity}/transfer/transfer_preimage`;
    const torProxy = clientConfig.torProxy;

    let socksAgent = undefined;

    if (torProxy) {
        socksAgent = { httpAgent: new SocksProxyAgent(torProxy) };
    }
    
    let response = await axios.post(url, transferPreimageRequestPayload, socksAgent);

    return { preimage: response?.data?.preimage };
}

const getPaymentHash = async (clientConfig, batchId) => {

    const url = `${clientConfig.statechainEntity}/transfer/paymenthash/${batchId}`;

    const torProxy = clientConfig.torProxy;

    let socksAgent = undefined;

    if (torProxy) {
        socksAgent = { httpAgent: new SocksProxyAgent(torProxy) };
    }

    try {
        const response = await axios.get(url, socksAgent);
        return response?.data?.hash;
    }
    catch (error) {
        if (error.response.status == 401) {
            return null;
        } else {
            throw new Error(`Failed to retrieve payment hash: ${JSON.stringify(error.response.data)}`);
        }
    }
}

module.exports = { createPreImage, confirmPendingInvoice, retrievePreImage, getPaymentHash };
